src/partykit/gateway.ts

```diff
@@ -0,0 +1,303 @@
+  }
+
+  async ready() {
+    return pkSockets.get(pkKey(this.pso)).once(async () => {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      this.party = new PartySocket(this.pso!);
+      // // needed to have openFn to be a stable reference
+
+      let exposedResolve: (value: boolean) => void;
+
+      const openFn = () => {
+        this.logger.Debug().Msg("party open");
+
+        // add our event listener
+        this.party?.addEventListener("message", async (event: MessageEvent<string>) => {
+          this.logger.Debug().Msg(`got message: ${event.data}`);
+          const enc = new TextEncoder();
+          this.messageResolve?.(enc.encode(event.data));
+          setTimeout(() => {
+            this.messagePromise = new Promise<Uint8Array>((resolve) => {
+              this.messageResolve = resolve;
+            });
+          }, 0);
+        });
+        this.logger.Debug().Msg("Resolving exposed resolve");
+        exposedResolve(true);
+      };
+
+      return await new Promise<boolean>((resolve) => {
+        exposedResolve = resolve;
+        this.party?.addEventListener("open", openFn);
+      });
+    });
+  }
+
+  async close(url: URI): Promise<bs.VoidResult> {
+    await this.ready();
+    this.logger.Debug().Msg("close");
+    // this.party?.close()
+    return Result.Ok(undefined);
+  }
+
+  async put(uri: URI, body: Uint8Array): Promise<Result<void>> {
+    this.logger.Debug().Msg(`about to put: ${uri.toString()}`);
+    await this.ready();
+    return exception2Result(async () => {
+      const { store } = getStore(uri, this.sthis, (...args) => args.join("/"));
+      switch (store) {
+        case "meta":
+          this.logger.Debug().Msg(`meta store put operation with body: ${new TextDecoder().decode(body)}`);
+          this.party?.send(new TextDecoder().decode(body));
+          break;
+        default:
+          await this.dataUpload(uri, body);
+      }
+    });
+  }
+
+  async dataUpload(uri: URI, bytes: Uint8Array) {
+    const key = uri.getParam("key");
+    if (!key) throw new Error("key not found");
+    const uploadUrl = pkCarURL(this.party, uri, key);
+    const response = await fetch(uploadUrl.toString(), { method: "PUT", body: bytes });
+    if (response.status === 404) {
+      throw new Error("Failure in uploading data!");
+    }
+  }
+
+  async subscribe(uri: URI, callback: (data: Uint8Array) => void): Promise<bs.VoidResult> {
+    this.logger.Debug().Msg(`about to subscribe: ${uri.toString()}`);
+    await this.ready();
+    this.logger.Debug().Msg(`subscribe ready: ${uri.toString()}`);
+    return exception2Result(async () => {
+      const store = uri.getParam("store");
+      switch (store) {
+        case "meta":
+          this.party?.addEventListener("message", async (event: MessageEvent<string>) => {
+            this.logger.Debug().Msg(`got message: ${event.data}`);
+            const enc = new TextEncoder();
+            callback(enc.encode(event.data));
```

Comment:
this is kind of a hack because intead of making a set of subscribers and calling them from the onmessage established in ready, I am just making a new on message subscriber for each one. there will be N = 1 typically